fork_choice_control/src/block_processor.rs

```diff
@@ -27,20 +35,80 @@ use types::{
     traits::{BeaconBlock as _, BeaconState as _, SignedBeaconBlock as _},
 };
 
+#[derive(Default)]
+pub struct TimingMetrics {
+    pub times: VecDeque<Duration>,
+    pub min: Duration,
+    pub max: Duration,
+    pub total: Duration,
+    pub count: usize,
+}
+
+impl TimingMetrics {
+    fn update(&mut self, duration: Duration) {
+        self.times.push_back(duration);
+        if self.times.len() > 100 {
+            if let Some(old) = self.times.pop_front() {
+                self.total -= old;
+            }
+        }
+        self.total += duration;
+        self.count += 1;
+        self.min = if self.count == 1 { duration } else { self.min.min(duration) };
+        self.max = self.max.max(duration);
+    }
+
+    fn average(&self) -> Duration {
+        if self.count > 0 {
+            self.total / self.count as u32
+        } else {
+            Duration::default()
+        }
+    }
+
+    fn median(&self) -> Duration {
+        let mut sorted = self.times.clone().into_iter().collect::<Vec<_>>();
+        sorted.sort();
+        sorted.get(sorted.len() / 2).cloned().unwrap_or_default()
+    }
+}
```

Comment:
1. Bug with `count`: The `count` variable is incremented but never decremented when an element is removed from the front (`pop_front()`). You don't need to track `count` manually, as you can retrieve the length of `times` using `self.times.len()`, which is O(1).

2. Bug with `min` and `max`: The current `min` and `max` logic doesn't account for the fact that the oldest value being popped might be the current minimum or maximum. In such cases, `min` or `max` would hold stale values that no longer exist in the sequence. You should calculate these values from scratch every time an element is added or removed, ensuring they are always correct.

3. Bug with `median` for even-sized sequences: The current `median` logic only handles the case for an uneven number of elements. For even-sized sequences, the median should be the average of the two middle elements.

4. Bug with `min`, `max`, `average`, and `median` for empty sequences: In the case of an empty sequence, no meaningful value exists for `min`, `max`, `average`, or `median`. Return `Option<Duration>` for these methods, with `None` indicating the absence of data.

Reply:
My suggestion is to implement the structure as follows. I copy the values before returning in the `min` and `max` function solely for the sake of consistency; such that all return an `Option<Duration>` instead of the aforementioned two methods returning a reference instead. Since it's a single value overhead is minimal. 

```rust
#[derive(Default)]
pub struct TimingMetrics {
    pub times: VecDeque<Duration>,
    pub total: Duration,
    pub max_size: usize,
}

impl TimingMetrics {
    pub fn new(max_size: usize) -> Self {
        Self {
            times: VecDeque::with_capacity(max_size),
            total: Duration::default(),
            max_size,
        }
    }

    pub fn update(&mut self, duration: Duration) {
        if self.times.len() >= self.max_size {
            if let Some(old) = self.times.pop_front() {
                self.total -= old;
            }
        }
        self.times.push_back(duration);
        self.total += duration;
    }

    pub fn min(&self) -> Option<Duration> {
        self.times.iter().min().copied()
    }

    pub fn max(&self) -> Option<Duration> {
        self.times.iter().max().copied()
    }

    pub fn average(&self) -> Option<Duration> {
        (!self.times.is_empty()).then(|| self.total / self.times.len() as u32)
    }

    pub fn median(&self) -> Option<Duration> {
        let len = self.times.len();
        if len == 0 {
            return None;
        }

        let mut sorted: Vec<_> = self.times.iter().collect();
        sorted.sort();

        // If len is odd or even, this approach works for both
        let mid_idx = len / 2;
        Some((*sorted[mid_idx] + *sorted[len - 1 - mid_idx]) / 2);
    }
}
```